Yell at the user for combining ICs and restart

GiudGiud opened the pull request:
Right now the system lets you specify both. And the IC takes over. On timestep 0, one can see the initial condition appear, overriding the restart, and the system uses the ICs (which changes the result) 

This is unfortunate because users often have ICs in their input because they copied over the steady input to make a transient that restarts from a steady calculation. There generally is not much of a need for ICs in that restarted case unless:

- new fields are introduced and need to be initialized
- a steady solution is not the one we want to start from for the transient

Hence why I added a boolean to work around the error

refs #21423 
Does not close it since no error with just this PR

GiudGiud committed:
Dont allow checkpoint recovery/restart and exodus restart at the same time, refs #21423

GiudGiud committed:
Dont allow checkpoint/exodus restart and initial conditions at the same time
Unless user says it s fine
 refs #21423

GiudGiud committed:
Add boolean authorizing ICs and restart to tests that are doing both
The output of these tests (gold file) includes both the ICs and the restart, which is odd in general, and sometimes what someone would want to do

GiudGiud committed:
Add boolean authorizing ICs and restart to tests in the modules that are doing both
 refs #21423
The output of these tests (gold file) includes both the ICs and the restart, which is odd in general, and sometimes what someone would want to do

Fix some issues in porous flow (missing number)
Update a deprecated FP, format inputs on save

GiudGiud commented:
Applications with tests affected (check when patch is up)
- [ ] BISON
- [ ] Griffin
- [ ] Pronghorn
- [ ] RELAP7
- [ ] Falcon

GiudGiud commented:
Would love an approval on the design of this before I start making patches for apps

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1748291/) on 6d62d99 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/25395/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1748309/) on 6d62d99 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>395233</th>
      <th colspan=4>#25395 6d62d9</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.66%</td>
      <td>85.66%</td>
      <td>-0.00%</td>
      <td>85.71%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>90538</td>
      <td>90545</td>
      <td>+7</td>
      <td>12</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15154</td>
      <td>15160</td>
      <td>+6</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25395/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25395/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/25395/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/25395/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/25395/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/25395/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/25395/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/25395/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/25395/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/25395/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/25395/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/25395/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/25395/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/25395/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/25395/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/25395/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/25395/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/25395/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/25395/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/25395/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/25395/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/25395/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/25395/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/25395/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/25395/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/25395/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/25395/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/25395/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/25395/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/25395/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 85.71% is less than the suggested 90.0%



This comment will be updated on new commits.

joshuahansel commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -2966,6 +2972,22 @@ FEProblemBase::addInitialCondition(const std::string & ic_name,
+      restart_method = "an Exodus restart";
+    if (!restart_method.empty())
+      mooseError(
+          "Initial condition has been specified during ",
```
```suggestion
          "Initial conditions have been specified during ",
```

joshuahansel commented on test/tests/kernels/ode/coupled_ode_td_auxvar_ic_from_mesh.i:
```diff
@@ -4,68 +4,74 @@
+
+[Problem]
+  # There are initial conditions overwriting the restart on the nonlinear variables
+  # However this test is targetted at the auxiliary variable restart so it's ok
```
```suggestion
  # However this test is targeted at the auxiliary variable restart so it's ok
```

GiudGiud commented:
We can for Exodus I think. For checkpoint rn, I dont think so.
I ll add this

GiudGiud commented on test/tests/auxkernels/solution_aux/solution_aux_exodus_interp_restart2.i:
```diff
@@ -70,4 +70,7 @@
 [Problem]
   restart_file_base = solution_aux_exodus_interp_restart1_out_cp/0005
+  # There are initial conditions overwriting the restart on the nonlinear variables
+  # However this test is targetted at the auxiliary variable initialized from the solution uo so it's ok
```
```suggestion
  # However this test is targeted at the auxiliary variable initialized from the solution uo so it's ok
```

GiudGiud committed:
Apply suggestions from code review
- detect only conflict between exodus restart and IC when the variable is actually restarted
- output the variable name when there s a conflict with exodus and IC
- better spelling
Co-authored-by: joshuahansel <joshua.hansel@inl.gov>

lindsayad reviewed (commented):
This looks sensical to me

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1750718/) on b9fe5ef wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25395/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25395/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a1e20d32d7c0cf0e6d911a534c5ec640fc7306cd`

loganharbour reviewed (dismissed):
Would be better to make the default true, fix the apps without breaking them, and then change it to false

GiudGiud commented:
That s a great idea. I never seem to be able to think of these things

GiudGiud commented:
This should pass now. 

I agree with Josh that it s nice to be more restrictive and when we trigger this error. However:
- with checkpoint restart we are not selecting which variable to restarts, it should be all of them. So adding an initial condition is only a valid thing for a "NEW" variable, which is tough to know which one is new right now (in the future with flexible restarts I suppose it will be easier)

- with exodus restart, as soon as the variable is in the file we have the error. However some people restart only a few of the variables in the file. This is not in the spirit of restart but is very common (see flurry of failures in the apps). I think making people restart more things is good. so having the error in that situation is still a good thing

GiudGiud commented:
All patches are made. This needs to be merged before any patch can be merged.

joshuahansel commented:
> This means that if the spelling of a variable is different between exodus and the input file, we wont get the error. So not sure this is better but it s ok with me

I didn't know a variable could have a different spelling. How does this work - how do they get linked?

My suggestion was not necessarily to change the error behavior, but just to give additional useful information when possible. For instance, if people just blindly started copying and pasting `allow_initial_conditions_with_restart = true`. I'm ok with either error behavior, so you choose what's best.

GiudGiud commented:
> I didn't know a variable could have a different spelling. How does this work - how do they get linked?

I think they dont right now but it s a very artificial limitation
